consume.go

```diff
@@ -39,49 +39,130 @@ type consumeCmd struct {
 	poms          map[int32]sarama.PartitionOffsetManager
 }
 
-var offsetResume int64 = -3
+const (
+	maxOffset    int64 = 1<<63 - 1
+	offsetResume int64 = -3
+)
 
-type offset struct {
-	relative bool
-	start    int64
-	diff     int64
+type positionRange struct {
+	startAnchor, endAnchor anchor
+	diff                   anchorDiff
 }
 
-func (cmd *consumeCmd) resolveOffset(o offset, partition int32) (int64, error) {
-	if !o.relative {
-		return o.start, nil
+func (r positionRange) start() position {
+	return position{
+		anchor: r.startAnchor,
+		diff:   r.diff,
 	}
+}
 
-	var (
-		res int64
-		err error
-	)
+func (r positionRange) end() position {
+	return position{
+		anchor: r.endAnchor,
+		diff:   r.diff,
+	}
+}
+
+// position represents a position within the Kafka stream.
+// It holds an anchor (an absolute position specified as a
+// time stamp or an offset) and a relative position with
+// respect to that (specified as an offset delta or a time duration).
+type position struct {
```

Comment:
I have some struggle to understand the naming here. Position tends to mean a single position of an offset whereas here IIUC we are talking about a "range" of position because we have the anchor and another anchor which is the diff. It's not crystal clear to me. Can you enlighten on this?

Reply:
It's not a range - it's an absolute position that's specified as a absolute position and an offset from that position. For example `latest-100` is a position that is 100 messages before the latest available message. The offsets parsing code can't resolve that itself because it doesn't know the absolute offset of the latest available message, so we leave it in "symbolic" form in the `position` type.

Reply:
To understand more clearly, let's say the position is `latest-100` and the latest offset is 300.
The value will look like:
```go
position {
	anchor{false, 300, timeZeroValue},
	anchorDiff{false, 100, timeZeroValue},
}
```

Reply:
The actual value of the latest offset is irrelevant here because the parsing code has no way to know what the actual latest offset is (it's just parsing some text and has no access to Kafka).

For the `latest-100` case, the value will look like:
```
position {
    anchor{false, sarama.OffsetNewest, time.Time{}},
    anchorDiff{false, -100, time.Time{}},
}
```

Note the special `sarama.OffsetNewest` constant which represents the newest offset symbolically. There are two other special symbolic offset values (`sarama.OffsetOldest` and the locally defined `offsetResume`). Note that they're all negative so there's no clash with real absolute offsets which are all non-negative.

Reply:
Alright, got it! Last one: the `offsetResume` is just to resume where the consumer is and compute the diff from it, right?

Reply:
that's right, yes.